src/apps/testapps/testH3CellAreaExhaustive.c

```diff
@@ -0,0 +1,173 @@
+ *        of indexes.
+ *
+ *  usage: `testH3CellAreaExhaustive`
+ */
+
+#include <math.h>
+
+#include "h3Index.h"
+#include "test.h"
+#include "utility.h"
+
+/**
+ * Basic checks around the great circle distance between the centers of two
+ * neighboring cells. Tests positivity and commutativity.
+ *
+ * Tests the functions:
+ *     pointDistRads
+ *     pointDistKm
+ *     pointDistM
+ *
+ * @param  edge  H3 unidirectional edge denoting neighboring cells
+ */
+static void haversine_assert(H3Index edge) {
+    GeoCoord a, b;
+
+    H3_EXPORT(h3ToGeo)
+    (H3_EXPORT(getOriginH3IndexFromUnidirectionalEdge)(edge), &a);
+    H3_EXPORT(h3ToGeo)
+    (H3_EXPORT(getDestinationH3IndexFromUnidirectionalEdge)(edge), &b);
+
+    char pos[] = "distance between cell centers should be positive";
+    char comm[] = "pairwise cell distances should be commutative";
+
+    double ab, ba;
+
+    ab = H3_EXPORT(pointDistRads)(&a, &b);
+    ba = H3_EXPORT(pointDistRads)(&b, &a);
+    t_assert(ab > 0, pos);
+    t_assert(ab == ba, comm);
+
+    ab = H3_EXPORT(pointDistKm)(&a, &b);
+    ba = H3_EXPORT(pointDistKm)(&b, &a);
+    t_assert(ab > 0, pos);
+    t_assert(ab == ba, comm);
+
+    ab = H3_EXPORT(pointDistM)(&a, &b);
+    ba = H3_EXPORT(pointDistM)(&b, &a);
+    t_assert(ab > 0, pos);
+    t_assert(ab == ba, comm);
+}
+
+/**
+ * Tests positivity of edge length calculation for the functions:
+ *
+ *      exactEdgeLengthRads
+ *      exactEdgeLengthKm
+ *      exactEdgeLengthM
+ *
+ * @param  edge  edge to compute the length of
+ */
+static void edge_length_assert(H3Index edge) {
+    char msg[] = "edge has positive length";
+
+    t_assert(H3_EXPORT(exactEdgeLengthRads)(edge) > 0, msg);
+    t_assert(H3_EXPORT(exactEdgeLengthKm)(edge) > 0, msg);
+    t_assert(H3_EXPORT(exactEdgeLengthM)(edge) > 0, msg);
+}
+
+/**
+ * Test that cell area calculations are positive for the functions:
+ *
+ *      cellAreaRads2
+ *      cellAreaKm2
+ *      cellAreaM2
+ *
+ * @param  cell  cell to compute the area of
+ */
+static void cell_area_assert(H3Index cell) {
+    char msg[] = "cell has positive length";
+
+    t_assert(H3_EXPORT(cellAreaRads2)(cell) > 0, msg);
+    t_assert(H3_EXPORT(cellAreaKm2)(cell) > 0, msg);
+    t_assert(H3_EXPORT(cellAreaM2)(cell) > 0, msg);
+}
+
+/**
+ * Apply a cell area calculation function to every cell on the earth at a given
+ * resolution, and check that it sums up the total earth area.
+ *
+ * @param  res        resolution of the cells
+ * @param  cell_area  callback to compute area of each cell
+ * @param  target     expected earth area in some units
+ * @param  tol        error tolerance allowed between expected and actual
+ */
+static void earth_area_test(int res, double (*cell_area)(H3Index),
+                            double target, double tol) {
+    double area = mapSumAllCells_double(res, cell_area);
+
+    t_assert(fabs(area - target) < tol,
+             "sum of all cells should give earth area");
+}
+
```

Comment:
Is it possible to assert some other stats (e.g. standard deviation)? I am thinking about a failure case where one cell has the area of the entire Earth and every other cell is 0.

Reply:
I think this is now covered by the newer test that gets the cell area for the cells at each resolution containing `(0,0)`, and also that we ensure each cell area is **strictly** greater than zero.